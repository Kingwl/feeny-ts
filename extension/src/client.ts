import * as path from 'path';
import { workspace, ExtensionContext } from 'vscode';

import {
    LanguageClient,
    LanguageClientOptions,
    ServerOptions,
    TransportKind
} from 'vscode-languageclient/node';

let client: LanguageClient;

export function activate(context: ExtensionContext) {
    const serverModule = context.asAbsolutePath(path.join( 'dist', 'server.js'));
    const debugOptions = { execArgv: ['--nolazy', '--inspect=6009'] };

    const serverOptions: ServerOptions = {
        run: { module: serverModule, transport: TransportKind.ipc },
        debug: {
            module: serverModule,
            transport: TransportKind.ipc,
            options: debugOptions
        }
    };

    const clientOptions: LanguageClientOptions = {
        documentSelector: [{ scheme: 'file', language: 'feeny' }],
        synchronize: {
            fileEvents: workspace.createFileSystemWatcher('**/.feeny')
        }
    };

    client = new LanguageClient(
        'feeny',
        'Feeny language features',
        serverOptions,
        clientOptions
    );

    client.start();
}

export function deactivate(): Thenable<void> | undefined {
    if (!client) {
        return undefined;
    }
    return client.stop();
}
