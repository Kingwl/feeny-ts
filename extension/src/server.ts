import {
    createConnection,
    TextDocuments,
    DiagnosticSeverity,
    ProposedFeatures,
    InitializeParams,
    TextDocumentSyncKind,
    InitializeResult,
    Range
} from 'vscode-languageserver/node';

import { TextDocument } from 'vscode-languageserver-textdocument';

import { createLanguageService, TextSpan } from 'feeny-ts'

let connection = createConnection(ProposedFeatures.all);

let documents: TextDocuments<TextDocument> = new TextDocuments(TextDocument);

connection.onInitialize((params: InitializeParams) => {
    const result: InitializeResult = {
        capabilities: {
            textDocumentSync: TextDocumentSyncKind.Full,
            definitionProvider: true,
        }
    };
    return result;
});

let server: ReturnType<typeof createLanguageService> | undefined

function textSpanToRange (doc: TextDocument, span: TextSpan) {
    const startPosition = doc.positionAt(span.pos)
    const endPosition = doc.positionAt(span.end)

    return Range.create(
        startPosition,
        endPosition
    )
}

function createLsForDoc (doc: TextDocument) {
    server = createLanguageService(doc.getText())
    const diagnostics = server.getDiagnostics()
    connection.sendDiagnostics({
        uri: doc.uri,
        diagnostics: diagnostics.map(d => {
            return {
                range: textSpanToRange(doc, d.span),
                severity: DiagnosticSeverity.Error,
                message: d.message
            }
        })
    })
}

documents.onDidOpen(doc => {
    createLsForDoc(doc.document)
})

documents.onDidChangeContent(change => {
    createLsForDoc(change.document)
})

connection.onDidChangeWatchedFiles(_change => {
    connection.console.log('We received a file change event');
});

connection.onDefinition(async (params) => {
    if (!server) {
        return undefined;
    }

    const uri = params.textDocument.uri;
    const doc = documents.get(uri);
    const pos = doc.offsetAt(params.position);
    const result = server.goToDefinition(pos);
    if (!result) {
        return undefined;
    }

    const range = textSpanToRange(doc, result)
    result.pos
    return [
        {
            uri,
            range
        }
    ];
})

documents.listen(connection);

connection.listen();
