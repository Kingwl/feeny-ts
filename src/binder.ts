import { ASTNode, SourceFile } from "./types";

class Symbol {
    constructor(
        public name: string,
        public declaration: ASTNode
    ) {

    }
}

class SymbolTable {

}

export function createBinder(file: SourceFile) {

    function bind () {

    }
}
