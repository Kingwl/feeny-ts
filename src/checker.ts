import { AllDeclaration, Declaration, FunctionBase, ParamsAndReturnType, Symbol, SymbolFlag } from ".";
import { VariableStatement, MethodSlotSignatureDeclaration, ObjectSlot, ObjectSlotSignature, TypeNode, VariableSlotSignatureDeclaration, ArraysExpression, ASTNode, BreakExpression, ContinueExpression, Expression, ExpressionStatement, FunctionStatement, ParenExpression, PrintingExpression, FunctionCallExpression, FunctionExpression, IfExpression, MethodCallExpression, ObjectsExpression, SequenceOfStatements, SlotAssignmentExpression, SlotLookupExpression, SourceFile, SyntaxKind, ThisExpression, VariableAssignmentExpression, WhileExpression, BinaryShorthand, GetShorthand, SetShorthand, MethodSlot, VariableSlot, TypeDefDeclaration, ArraysTypeNode, TypeReferenceTypeNode, ParameterDeclaration, VariableReferenceExpression } from "./types";
import { assertKind, first, frontAndTail, isDeclaration, isDef, isExpression } from "./utils";
import { forEachChild } from './visitor'

enum TypeKind {
    Unknown,
    Never,
    Null,
    Integer,
    String,
    Object,
    Function,
    Union
}

interface Type {
    kind: TypeKind

    _debugKind?: string
}

interface UnknownType extends Type {
    kind: TypeKind.Unknown
}

interface NeverType extends Type {
    kind: TypeKind.Never
}

interface NullType extends Type {
    kind: TypeKind.Null
}

interface IntegerType extends Type {
    kind: TypeKind.Integer
}

interface StringType extends Type {
    kind: TypeKind.String
}

interface ObjectType extends Type {
    kind: TypeKind.Object
    properties: Map<string, Type>
}

interface FunctionType extends Type {
    kind: TypeKind.Function
    thisType?: Type
    paramTypes: Type[]
    returnType: Type
}

interface UnionType extends Type {
    kind: TypeKind.Union
    types: Type[]
}

export function createChecker(file: SourceFile) {
    const unknownType = createUnknownType();
    const errorType = createNeverType();
    const neverType = createNeverType();
    const nullType = createNullType();
    const integerType = createIntegerType();
    const stringType = createStringType();

    const typeCheckCache = new Map<ASTNode, Type | undefined>();

    return {
        checkFile,
    }

    function checkFile () {
        check(file);

        return {
            check
        }
    }

    function check (node: ASTNode): Type | undefined {
        if (typeCheckCache.has(node)) {
            return typeCheckCache.get(node);
        }

        const type = checkWorker(node);
        typeCheckCache.set(node, type);
        return type;
    }

    function checkWorker (node: ASTNode) {
        if (isExpression(node)) {
            return checkExpression(node)
        }

        if (isDeclaration(node)) {
            return checkDeclaration(node)
        }

        if (node.kind === SyntaxKind.SequenceOfStatements) {
            return checkSequenceOfStatements(node as SequenceOfStatements);
        }

        forEachChild(node, check);
        return undefined;
    }

    function checkDeclaration (node: Declaration): Type {
        switch (node.kind) {
            case SyntaxKind.FunctionStatement:
                return checkFunctionStatement(node as FunctionStatement)
            case SyntaxKind.VariableStatement:
                return checkVariableStatement(node as VariableStatement)
            case SyntaxKind.TypeDefDeclaration:
                return checkTypeDefDeclaration(node as TypeDefDeclaration);
            case SyntaxKind.MethodSlot:
                return checkMethodSlot(node as MethodSlot);
            case SyntaxKind.VariableSlot:
                return checkVariableSlot(node as VariableSlot);
            case SyntaxKind.MethodSlotSignatureDeclaration:
                return checkMethodSlotSignature(node as MethodSlotSignatureDeclaration);
            case SyntaxKind.VariableSlotSignatureDeclaration:
                return checkVariableSlotSignature(node as VariableSlotSignatureDeclaration);
            case SyntaxKind.ParameterDeclaration:
                return checkParameterDeclaration(node as ParameterDeclaration);
            case SyntaxKind.ObjectsExpression:
                return checkObjectsExpression(node as ObjectsExpression);
            default:
                throw new Error(`Unknown declaration kind ${node}`)
        }
    }

    function checkExpression (node: Expression): Type {
        switch (node.kind) {
            case SyntaxKind.IntegerLiteralExpression:
                return integerType
            case SyntaxKind.StringLiteralExpression:
                return stringType
            case SyntaxKind.NullKeyword:
                return nullType;
            case SyntaxKind.PrintingExpression:
                assertKind<PrintingExpression>(node)
                return checkPrintingExpression(node);
            case SyntaxKind.ArraysExpression:
                assertKind<ArraysExpression>(node)
                return checkArraysExpression(node);
            case SyntaxKind.ObjectsExpression:
                assertKind<ObjectsExpression>(node)
                return checkObjectsExpression(node);
            case SyntaxKind.MethodCallExpression:
                assertKind<MethodCallExpression>(node)
                return checkMethodCallExpression(node);
            case SyntaxKind.SlotLookupExpression:
                assertKind<SlotLookupExpression>(node)
                return checkSlotLookupExpression(node);
            case SyntaxKind.SlotAssignmentExpression:
                assertKind<SlotAssignmentExpression>(node)
                return checkSlotAssignmentExpression(node);
            case SyntaxKind.FunctionCallExpression:
                assertKind<FunctionCallExpression>(node)
                return checkFunctionCallExpression(node);
            case SyntaxKind.VariableAssignmentExpression:
                assertKind<VariableAssignmentExpression>(node)
                return checkVariableAssignmentExpression(node);
            case SyntaxKind.VariableReferenceExpression:
                assertKind<VariableReferenceExpression>(node);
                return checkVariableReferenceExpression(node);
            case SyntaxKind.IfExpression:
                assertKind<IfExpression>(node)
                return checkIfExpression(node);
            case SyntaxKind.WhileExpression:
                assertKind<WhileExpression>(node)
                return checkWhileExpression(node);
            case SyntaxKind.ThisExpression:
                assertKind<ThisExpression>(node)
                return checkThisExpression(node);
            case SyntaxKind.ParenExpression:
                assertKind<ParenExpression>(node)
                return checkParenExpression(node);
            case SyntaxKind.FunctionExpression:
                assertKind<FunctionExpression>(node)
                return checkFunctionExpression(node);
            case SyntaxKind.BreakExpression:
            case SyntaxKind.ContinueExpression:
                assertKind<BreakExpression | ContinueExpression>(node);
                return checkBreakOrContinueStatement(node);
            case SyntaxKind.BinaryShorthand:
                assertKind<BinaryShorthand>(node);
                return checkBinaryShorthand(node);
            case SyntaxKind.GetShorthand:
                assertKind<GetShorthand>(node);
                return checkGetShorthand(node);
            case SyntaxKind.SetShorthand:
                assertKind<SetShorthand>(node);
                return checkSetShorthand(node);
            default:
                forEachChild(node, check);
                return errorType;
        }
    }

    function isRelatedTo(source: Type, target: Type) {
        return true
    }

    function checkVariableReferenceExpression(node: VariableReferenceExpression): Type {
        const symbol = resolveName(node.name.text, node, SymbolFlag.Value);
        if (!symbol) {
            return unknownType
        }
        return checkDeclaration(symbol.declaration) 
    }

    function checkVariableSlotSignatureOrMethodSlotSignature(node: ObjectSlotSignature): Type {
        switch (node.kind) {
            case SyntaxKind.VariableSlotSignatureDeclaration:
                return checkVariableSlotSignature(node as VariableSlotSignatureDeclaration);
            case SyntaxKind.MethodSlotSignatureDeclaration:
                return checkMethodSlotSignature(node as MethodSlotSignatureDeclaration);
            default:
                throw new Error(`Unknown kind ${node.__debugKind}`)
        }
    }

    function checkVariableSlotOrMethodSlot(node: ObjectSlot) {
        switch (node.kind) {
            case SyntaxKind.VariableSlot:
                return checkVariableSlot(node as VariableSlot);
            case SyntaxKind.MethodSlot:
                return checkMethodSlot(node as MethodSlot);
            default:
                throw new Error(`Unknown kind ${node.__debugKind}`)
        }
    }

    function checkTypeNode(node: TypeNode): Type {
        switch (node.kind) {
            case SyntaxKind.NullTypeNode:
                return nullType;
            case SyntaxKind.IntegerTypeNode:
                return integerType;
            case SyntaxKind.ArraysTypeNode:
                return checkArraysTypeNode(node as ArraysTypeNode);
            case SyntaxKind.TypeReferenceTypeNode:
                return checkTypeReferenceTypeNode(node as TypeReferenceTypeNode);
            default:
                throw new Error(`Unknown kind ${node.__debugKind}`)
        }
    }

    function resolveName (name: string, location: ASTNode, flag: SymbolFlag): Symbol | undefined {
        let current: ASTNode | undefined = location
        while (current) {
            const matched = current.locals?.get(name);
            if (isDef(matched) && matched.flags & flag) {
                return matched;
            }

            current = current.parent
        }
        return undefined;
    }

    function checkTypeReferenceTypeNode(node: TypeReferenceTypeNode): Type {
        return errorType
    }

    function checkArraysTypeNode(node: ArraysTypeNode): Type {
        const type = checkTypeNode(node.type);
        return createArrayType(type)
    }

    function checkMethodSlot(node: MethodSlot) {
        const type = checkParamsAndReturnType(node)
        check(node.body);
        return type;
    }
    
    function checkVariableSlot(node: VariableSlot) {
        if (node.type) {
            return checkTypeNode(node.type)
        }

        return unknownType
    }

    function checkVariableSlotSignature(node: VariableSlotSignatureDeclaration): Type {
        return checkTypeNode(node.type)
    }

    function checkMethodSlotSignature(node: MethodSlotSignatureDeclaration): Type {
        return checkParamsAndReturnType(node);
    }

    function checkTypeDefDeclaration(node: TypeDefDeclaration) {
        const properties = new Map<string, Type>();
        node.slots.forEach(slot => {
            const slotType = checkVariableSlotSignatureOrMethodSlotSignature(slot);
            properties.set(slot.name.text, slotType);
        })
        return createObjectType(properties);
    }

    function checkVariableAssignmentExpression(node: VariableAssignmentExpression) {
        const target = checkExpression(node.expression);
        const type = checkExpression(node.value)
        isRelatedTo(type, target);
        return type;
    }
    
    function checkIfExpression(node: IfExpression) {
        checkExpression(node.condition);
        
        const thenType = checkSequenceOfStatementsOrExpressionStatement(node.thenStatement);
        const elseType = node.elseStatement ? checkSequenceOfStatementsOrExpressionStatement(node.elseStatement) : nullType;

        const type = createUnionType([
            thenType,
            elseType
        ])
        return type
    }
    
    function checkWhileExpression(node: WhileExpression) {
        checkExpression(node.condition);
        check(node.body);
        return nullType
    }

    function checkBreakOrContinueStatement(node: BreakExpression | ContinueExpression) {
        return neverType
    }

    function checkParameterDeclaration(node: ParameterDeclaration) {
        if (node.type) {
            checkTypeNode(node.type)
        }
        return unknownType
    }

    function checkParamsAndReturnType(node: ParamsAndReturnType) {
        const paramTypes: Type[] = [];
        node.params.forEach(param => {
            const paramType = checkParameterDeclaration(param)
            paramTypes.push(paramType)
        })
        const returnType = node.type ? checkTypeNode(node.type) : unknownType;
        return createFunctionType(undefined, paramTypes, returnType);
    }

    function checkParenExpression(node: ParenExpression) {
        return checkExpression(node)
    }

    function checkThisExpression(node: ThisExpression) {
        return errorType
    }

    function checkFunctionCallExpression(node: FunctionCallExpression) {
        return errorType
    }

    function checkSlotAssignmentExpression(node: SlotAssignmentExpression) {
        const targetType = checkExpression(node.expression);
        const valueType = checkExpression(node.value);
        isRelatedTo(valueType, targetType);
        return valueType
    }

    function checkSlotLookupExpression(node: SlotLookupExpression) {
        return errorType
    }

    function checkMethodCallExpression(node: MethodCallExpression) {
        return errorType
    }

    function checkObjectsExpression(node: ObjectsExpression) {
        const properties = new Map<string, Type>();
        node.slots.forEach(slot => {
            const slotType = checkVariableSlotOrMethodSlot(slot);
            properties.set(slot.name.text, slotType);
        })
        return createObjectType(properties);
    }

    function checkArraysExpression(node: ArraysExpression) {
        check(node.length);

        const itemType = node.defaultValue ? checkExpression(node.defaultValue) : neverType;
        return createArrayType(itemType);
    }

    function checkPrintingExpression (node: PrintingExpression) {
        forEachChild(node, check);
        return nullType
    }

    function checkFunctionStatement(node: FunctionStatement): Type {
        const type = checkParamsAndReturnType(node)
        check(node.body);
        return type;
    }

    function checkFunctionExpression(node: FunctionExpression) {
        const type = checkParamsAndReturnType(node)
        check(node.body);
        return type;
    }

    function checkVariableStatement(node: VariableStatement): Type {
        if (node.type) {
            return checkTypeNode(node.type)
        }
        return unknownType
    }

    function checkExpressionStatement(node: ExpressionStatement) {
        return checkExpression(node.expression);
    }

    function checkSequenceOfStatementsOrExpressionStatement(node: SequenceOfStatements<true> | ExpressionStatement): Type
    function checkSequenceOfStatementsOrExpressionStatement(node: SequenceOfStatements<boolean> | ExpressionStatement): Type | undefined
    function checkSequenceOfStatementsOrExpressionStatement<T extends boolean>(node: SequenceOfStatements<T> | ExpressionStatement): Type | undefined {
        if (node.kind === SyntaxKind.SequenceOfStatements) {
            return checkSequenceOfStatements(node);
        } else {
            return checkExpressionStatement(node);
        }
    }

    function checkSequenceOfStatements(node: SequenceOfStatements<true>): Type
    function checkSequenceOfStatements(node: SequenceOfStatements): Type | undefined
    function checkSequenceOfStatements<T extends boolean>(node: SequenceOfStatements<T>): Type | undefined {
        if (!node.isExpression) {
            forEachChild(node, check);
            return undefined;
        }

        const [ front, tail ] = frontAndTail(node.statements)
        front.forEach(check);

        if (tail.kind !== SyntaxKind.ExpressionStatement) {
            return errorType
        }

        return checkExpressionStatement(tail as ExpressionStatement);
    }

    function checkBinaryShorthand(node: BinaryShorthand): Type {
        return errorType
    }
    
    function checkGetShorthand(node: GetShorthand): Type {
        return errorType
    }
    
    function checkSetShorthand(node: SetShorthand): Type {
        return errorType
    }

    function createUnknownType () {
        const type: UnknownType = {
            kind: TypeKind.Unknown
        }
        setupTypeDebugInfo(type);
        return type
    }
    
    function createNeverType () {
        const type: NeverType = {
            kind: TypeKind.Never
        }
        setupTypeDebugInfo(type);
        return type
    }
    
    function createNullType () {
        const type: NullType = {
            kind: TypeKind.Null
        }
        setupTypeDebugInfo(type);
        return type
    }
    
    function createIntegerType () {
        const type: IntegerType = {
            kind: TypeKind.Integer
        }
        setupTypeDebugInfo(type);
        return type
    }
    
    function createStringType () {
        const type: StringType = {
            kind: TypeKind.String
        }
        setupTypeDebugInfo(type);
        return type
    }
    
    function createObjectType (properties: Map<string, Type>) {
        const type: ObjectType = {
            kind: TypeKind.Object,
            properties
        }
        setupTypeDebugInfo(type);
        return type
    }
    
    function createFunctionType (thisType: Type | undefined, paramTypes: Type[], returnType: Type) {
        const type: FunctionType = {
            kind: TypeKind.Function,
            thisType,
            paramTypes,
            returnType
        }
        setupTypeDebugInfo(type);
        return type
    }
    
    function createUnionType (types: Type[]): Type {
        if (!types.length) {
            return neverType
        }

        if (types.length === 1) {
            return first(types)
        }
    
        const type: UnionType = {
            kind: TypeKind.Union,
            types
        }
        setupTypeDebugInfo(type);
        return type
    }

    function setupTypeDebugInfo (type: Type) {
        Object.defineProperty(type, '_debugKind', {
            get () {
                return TypeKind[type.kind]
            }
        })
    }

    function createArrayType (itemType: Type) {
        const properties = new Map<string, Type>();
        const arrayType = createObjectType(properties);

        const getMethod = createGetMethod();
        const setMethod = createSetMethod();
        const lengthMethod = createLengthMethod();

        properties.set("get", getMethod);
        properties.set("set", setMethod);
        properties.set("length", lengthMethod);
        return arrayType

        function createGetMethod () {
            return createFunctionType(
                arrayType,
                [integerType],
                itemType
            )
        }

        function createSetMethod () {
            return createFunctionType(
                arrayType,
                [integerType, itemType],
                nullType
            )
        }

        function createLengthMethod () {
            return createFunctionType(
                arrayType,
                [],
                integerType
            );
        }
    }
}
