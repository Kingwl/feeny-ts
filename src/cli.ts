import * as fs from 'fs';
import * as path from 'path';
import { createParser, createInterpreter } from './index';
import { Context } from './interpreter/types';

function main() {
  if (process.argv.length < 3) {
    console.error('Input missing.');
    return;
  }

  const file = process.argv[2];
  if (!file) {
    console.error(`Invalid input '${file}'.`);
    return;
  }

  const filePath = path.resolve(file);
  if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
    console.error(`File '${filePath}' not existed.`);
    return;
  }

  const content = fs.readFileSync(filePath).toString();
  if (!content) {
    console.error(`File '${filePath}' is empty.`);
    return;
  }

  const parser = createParser(content);
  const sourceFile = parser.parseSourceFile();
  const context: Context = {
    stdout: text => process.stdout.write(text, 'utf-8')
  }
  const interpreter = createInterpreter(sourceFile, context);
  interpreter.evaluate();
}

main();
