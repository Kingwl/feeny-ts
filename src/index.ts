export * from './types';
export * from './scanner';
export * from './parser';
export * from './factory';
export * from './binder';
export * from './visitor';
export * from './checker';
export * from './interpreter/interpreter';
