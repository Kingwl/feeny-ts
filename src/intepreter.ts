import { ASTNode, LocalStatement, SyntaxKind } from ".";
import { SequenceOfStatements, SourceFile, Statement, TopLevelStatement } from "./types";

enum ValueType {
    Null,
    Array,
    Object,
    Function,
    Method,
    Integer,
}

abstract class BaseValue {
    abstract get type(): ValueType;

    isNull(): this is NullValue {
        return false
    }

    isArray(): this is ArrayValue {
        return false;
    }

    isObject(): this is ObjectValue {
        return false;
    }

    isFunction(): this is FunctionValue {
        return false;
    }

    isMethod(): this is MethodValue {
        return false;
    }

    isInteger(): this is IntegerValue {
        return false
    }
}

class NullValue extends BaseValue {
    get type () { return ValueType.Null }

    isNull (): true {
        return true
    }
}

class ArrayValue extends BaseValue {
    get type () { return ValueType.Array }

    isArray(): true {
        return true;
    }
}

class ObjectValue extends BaseValue {
    get type () { return ValueType.Object }

    isObject(): true {
        return true
    }
}

class FunctionValue extends BaseValue {
    get type () { return ValueType.Function }

    isFunction(): true {
        return true
    }
}

class MethodValue extends BaseValue {
    get type () { return ValueType.Method }

    isMethod(): true {
        return true
    }
}

class IntegerValue extends BaseValue {
    get type () { return ValueType.Integer }

    isInteger(): true {
        return true;
    }
}

type VarValues = IntegerValue | ArrayValue | ObjectValue | NullValue
type CodeValues = FunctionValue | MethodValue
type AllValues = VarValues | CodeValues

function isVarValues (value: BaseValue): value is VarValues {
    return value.isInteger() || value.isNull() || value.isObject() || value.isArray();
}

function isCodeValues (value: BaseValue): value is CodeValues {
    return value.isFunction() || value.isMethod()
}

class Environment {
    private varValues = new Map<string, VarValues>()
    private codeValues = new Map<string, CodeValues>();

    hasBinding(name: string) {
        return this.varValues.has(name) || this.codeValues.has(name) 
    }

    addBinding(name: string, value: BaseValue) {
        if (isVarValues(value)) {
            this.varValues.set(name, value)
        } else if (isCodeValues(value)) {
            this.codeValues.set(name, value)
        } else {
            throw new Error("Invalid value type")
        }
    }

    getBinding(name: string) {
        return this.varValues.get(name) ?? this.codeValues.get(name)
    }
}

export function createIntepreter(file: SourceFile) {
    const globalEnv = new Environment()
    const envStack = [globalEnv];

    evaluate(file);

    function evaluate(node: ASTNode) {
        switch (node.kind) {
            case SyntaxKind.SourceFile:

        }
    }
}