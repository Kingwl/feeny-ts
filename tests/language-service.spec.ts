import { createLanguageService } from '../src'

describe('Language service', () => {
    const code = 
`
typedef ab:
    var a: integer
    var b: integer

defn foo() -> ab:
    object:
        var a = 1
        var b = 2
var bar = foo()
bar.a
bar.b
`     

    const posList = [
        code.indexOf(`bar.`),
        code.indexOf(`bar.a`) + 'bar.'.length,
        code.indexOf(`bar.b`) + 'bar.'.length,
        code.indexOf(`= foo()`) + "= ".length,
    ]
    const ls = createLanguageService(code)
    posList.forEach(pos => {
        const decl = ls.goToDefinition(pos)
        it(`Definition of ${pos}`, () => {
            expect(decl?.__debugText).toMatchSnapshot()
        })
    })
})