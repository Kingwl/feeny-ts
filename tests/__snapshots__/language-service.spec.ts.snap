// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language service Definition of 128 1`] = `
"defn foo() -> ab:
    object:
        var a = 1
        var b = 2"
`;

exports[`Language service Definition of 134 1`] = `"var bar = foo()"`;

exports[`Language service Definition of 138 1`] = `"var a: integer"`;

exports[`Language service Definition of 144 1`] = `"var b: integer"`;
